runtime/server/api.go

```diff
@@ -78,3 +84,204 @@ func (s *Server) apiHandler(w http.ResponseWriter, req *http.Request) error {
 
 	return nil
 }
+
+func (s *Server) combinedOpenAPISpec(w http.ResponseWriter, req *http.Request) error {
+	// Parse path parameters
+	ctx := req.Context()
+	instanceID := req.PathValue("instance_id")
+
+	// Add observability attributes
+	observability.AddRequestAttributes(ctx)
+	s.addInstanceRequestAttributes(ctx, instanceID)
+
+	// Only GET request is allowed
+	if req.Method != http.MethodGet {
+		return httputil.Error(http.StatusMethodNotAllowed, fmt.Errorf("GET only"))
+	}
+
+	// Check if user has access to query for API data
+	if !auth.GetClaims(ctx).CanInstance(instanceID, auth.ReadAPI) {
+		return httputil.Errorf(http.StatusForbidden, "does not have access to custom APIs")
+	}
+
+	apis := make(map[string]*runtimev1.API)
+
+	// Get all built-in APIs
+	for name, api := range runtime.BuiltinAPIs {
+		apis[name] = api
+	}
+
+	// Get all custom APIs
+	ctrl, err := s.runtime.Controller(ctx, instanceID)
+	if err != nil {
+		return httputil.Error(http.StatusBadRequest, err)
+	}
+
+	list, err := ctrl.List(ctx, runtime.ResourceKindAPI, "", false)
+	if err != nil {
+		return err
+	}
+
+	for _, res := range list {
+		apis[res.Meta.Name.Name] = res.GetApi()
+	}
+
+	// Generate the OpenAPI spec
+	combinedAPISpec, err := s.generateOpenAPISpec(instanceID, apis)
+	if err != nil {
+		return httputil.Error(http.StatusInternalServerError, err)
+	}
+
+	// Check accepted format of the OpenAPI spec: JSON or YAML
+	accept := req.Header.Get("Accept")
+
+	// YAML
+	if accept == "application/yaml" || accept == "application/x-yaml" {
+		data, err := yaml.Marshal(combinedAPISpec)
+		if err != nil {
+			return httputil.Error(http.StatusInternalServerError, err)
+		}
+
+		w.Header().Set("Content-Type", "application/yaml")
+		_, err = w.Write(data)
+		if err != nil {
+			return httputil.Error(http.StatusInternalServerError, err)
+		}
+
+		return nil
+	}
+
+	// JSON
+	data, err := combinedAPISpec.MarshalJSON()
+	if err != nil {
+		return httputil.Error(http.StatusInternalServerError, err)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(data)
+	if err != nil {
+		return httputil.Error(http.StatusInternalServerError, err)
+	}
+
+	return nil
+}
+
+func (s *Server) generateOpenAPISpec(instanceID string, apis map[string]*runtimev1.API) (*spec.Swagger, error) {
+	baseSpec := &spec.Swagger{
+		SwaggerProps: spec.SwaggerProps{
+			Swagger:  "2.0",
+			Produces: []string{"application/json"},
+			Info: &spec.Info{
+				InfoProps: spec.InfoProps{
+					Title:   "Rill custom API of instance " + instanceID,
```

Comment:
Could we pull the org/project name from the instance annotations and incorporate those here? And if not found, default to just `Rill project API` or something like that?